src/compiler/checker.ts

```diff
@@ -17005,7 +17005,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         const typeVariables: TypeVariable[] = [];
         // First collect a list of the type variables occurring in constraining intersections.
         for (const type of types) {
-            if (getObjectFlags(type) & ObjectFlags.IsConstrainedTypeVariable) {
+            if (type.flags & TypeFlags.Intersection && getObjectFlags(type) & ObjectFlags.IsConstrainedTypeVariable) {

```

Comment:
The problem is that `ObjectFlags` reuse some values as they are reserved for certain types, like unions, intersections, substitutions, etc. So the same value is being reused between `ObjectFlags.IdenticalBaseTypeExists` and `ObjectFlags.IsConstrainedTypeVariable`. That created a problem here since the code in this branch assumes that it's dealing with an intersection based on this value - and yet it encountered an object with `ObjectFlags.IdenticalBaseTypeExists`

The union creation here starts with three types:
```
['Element', 'string', 'Capitalize<string>']
```

The `Element` is the one that has `ObjectFlags.IdenticalBaseTypeExists` while `Capitalize<string>` is the one with `StringMapping|IncludesConstrainedTypeVariable` (where `TypeFlags.IncludesConstrainedTypeVariable` is defined as `= StringMapping`).

This `Capitalize<string>` gets eliminated from this union since `string` is its supertype - but it's the one that adds the `TypeFlags.IncludesConstrainedTypeVariable` to the `includes` returned by `addTypesToUnion` (that happens based on the `TypeFlags.IncludesMask`).

Reply:
Eh, no, that's not helpful; it'd just be preserved if overlapping.